[plan cache] panic on `explain for connection`

## Bug Report

I wrote a customer sysbench test, the lua script file is named issue.lua, and the content is:

```lua
#!/usr/bin/env sysbench

function thread_init()
    drv = sysbench.sql.driver()
    con = drv:connect()
    con:query([[
        prepare stmt from "
            select * from t1, t2
            where t1.a = t2.a
            and t2.a = ?
            and t2.b = 'b'
            and t2.c = 'c'
        "
    ]])
end

function thread_done()
    con:disconnect()
end

function event()
    con:query("set @a = 1")
    con:query("execute stmt using @a")
end

function prepare()
    local drv = sysbench.sql.driver()
    local con = drv:connect()

    con:query("DROP TABLE IF EXISTS t1")
    con:query([[
        CREATE TABLE t1 (
            a     BIGINT,
            b     VARCHAR(40),
            col3  VARCHAR(40) DEFAULT NULL,
            col4  VARCHAR(40) DEFAULT NULL,
            col5  VARCHAR(40) DEFAULT NULL,
            col6  VARCHAR(40) DEFAULT NULL,
            col7  VARCHAR(40) DEFAULT NULL,
            col8  VARCHAR(40) DEFAULT NULL,
            col9  VARCHAR(40) DEFAULT NULL,
            col10 VARCHAR(40) DEFAULT NULL,
            PRIMARY KEY (a, b)
        )
    ]])

    con:query([[INSERT INTO t1(a, b) VALUES (1, "1")]])
    con:query([[INSERT INTO t1(a, b) VALUES (1, "2")]])
    con:query([[INSERT INTO t1(a, b) VALUES (2, "1")]])
    con:query([[INSERT INTO t1(a, b) VALUES (2, "2")]])
    con:query([[INSERT INTO t1(a, b) VALUES (3, "1")]])
    con:query([[INSERT INTO t1(a, b) VALUES (3, "2")]])

    con:query("DROP TABLE IF EXISTS t2")
    con:query([[
        CREATE TABLE t2 (
            a     BIGINT,
            b     VARCHAR(40),
            c     VARCHAR(40),
            col4  VARCHAR(40) DEFAULT NULL,
            col5  VARCHAR(40) DEFAULT NULL,
            col6  VARCHAR(40) DEFAULT NULL,
            col7  VARCHAR(40) DEFAULT NULL,
            col8  VARCHAR(40) DEFAULT NULL,
            col9  VARCHAR(40) DEFAULT NULL,
            col10 VARCHAR(40) DEFAULT NULL,
            UNIQUE KEY idx_a (a)
        )
    ]])

    con:query([[INSERT INTO t2(a, b, c) VALUES(1, "b", "c")]])
    con:query([[INSERT INTO t2(a, b, c) VALUES(2, "b", "c")]])
    con:query([[INSERT INTO t2(a, b, c) VALUES(3, "b", "c")]])
end

function cleanup()
    local drv = sysbench.sql.driver()
    local con = drv:connect()
    con:query("DROP TABLE IF EXISTS t1")
    con:query("DROP TABLE IF EXISTS t2")
end
```

**NOTE**: the content is different from it described in #16817.

### 1. What did you do?

- step 1: step 1: start a tidb-server with mocktikv and **plan cache enabled**

- step 2: prepare table and data via the sysbench test:

```sh
sysbench --report-interval=1 --time=300 --threads=4 --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-db=test --db-driver=mysql issue prepare
```

- step 3: run the sysbench test:

```sh
sysbench --report-interval=1 --time=300 --threads=4 --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-db=test --db-driver=mysql issue run
```

- step 4: got one of the connection id through log file, open a MySQL client and run `explain for connection xx` to see the execution plan on that connection, I got an error in the return result:

```sql
TiDB(root@127.0.0.1:test) > explain for connection 2;
ERROR 1105 (HY000): runtime error: index out of range [0] with length 0
```

and the panic stack in tidb log file is:

```
[2020/04/25 17:24:11.449 +08:00] [ERROR] [adapter.go:123] ["execute sql panic"] [conn=6] [sql="explain for connection 2"] [stack="github.com/pingcap/tidb/executor.(*recordSet).Next.func1\n\t/Users/jianzhang.zj/Code/tidb/executor/adapter.go:123\nruntime.gopanic\n\t/opt/goroot/src/runtime/panic.go:679\nruntime.goPanicIndex\n\t/opt/goroot/src/runtime/panic.go:75\ngithub.com/pingcap/tidb/expression.(*ParamMarker).GetUserVar\n\t/Users/jianzhang.zj/Code/tidb/expression/constant.go:80\ngithub.com/pingcap/tidb/expression.(*Constant).getLazyDatum\n\t/Users/jianzhang.zj/Code/tidb/expression/constant.go:186\ngithub.com/pingcap/tidb/expression.(*Constant).EvalString\n\t/Users/jianzhang.zj/Code/tidb/expression/constant.go:264\ngithub.com/pingcap/tidb/expression.(*builtinCastStringAsRealSig).evalReal\n\t/Users/jianzhang.zj/Code/tidb/expression/builtin_cast.go:1158\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).EvalReal\n\t/Users/jianzhang.zj/Code/tidb/expression/scalar_function.go:354\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).Eval\n\t/Users/jianzhang.zj/Code/tidb/expression/scalar_function.go:323\ngithub.com/pingcap/tidb/expression.(*Constant).getLazyDatum\n\t/Users/jianzhang.zj/Code/tidb/expression/constant.go:190\ngithub.com/pingcap/tidb/expression.(*Constant).Eval\n\t/Users/jianzhang.zj/Code/tidb/expression/constant.go:199\ngithub.com/pingcap/tidb/expression.(*Constant).ExplainInfo\n\t/Users/jianzhang.zj/Code/tidb/expression/explain.go:65\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).explainInfo\n\t/Users/jianzhang.zj/Code/tidb/expression/explain.go:38\ngithub.com/pingcap/tidb/expression.(*ScalarFunction).ExplainInfo\n\t/Users/jianzhang.zj/Code/tidb/expression/explain.go:28\ngithub.com/pingcap/tidb/expression.sortedExplainExpressionList\n\t/Users/jianzhang.zj/Code/tidb/expression/explain.go:119\ngithub.com/pingcap/tidb/expression.SortedExplainExpressionList\n\t/Users/jianzhang.zj/Code/tidb/expression/explain.go:109\ngithub.com/pingcap/tidb/planner/core.(*PhysicalSelection).ExplainInfo\n\t/Users/jianzhang.zj/Code/tidb/planner/core/explain.go:282\ngithub.com/pingcap/tidb/planner/core.(*Explain).prepareOperatorInfo\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:953\ngithub.com/pingcap/tidb/planner/core.(*Explain).explainPlanInRowFormat\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:839\ngithub.com/pingcap/tidb/planner/core.(*Explain).explainPlanInRowFormat\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:903\ngithub.com/pingcap/tidb/planner/core.(*Explain).explainPlanInRowFormat\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:886\ngithub.com/pingcap/tidb/planner/core.(*Explain).explainPlanInRowFormat\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:886\ngithub.com/pingcap/tidb/planner/core.(*Explain).explainPlanInRowFormat\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:928\ngithub.com/pingcap/tidb/planner/core.(*Explain).RenderResult\n\t/Users/jianzhang.zj/Code/tidb/planner/core/common_plans.go:821\ngithub.com/pingcap/tidb/executor.(*ExplainExec).generateExplainInfo\n\t/Users/jianzhang.zj/Code/tidb/executor/explain.go:106\ngithub.com/pingcap/tidb/executor.(*ExplainExec).Next\n\t/Users/jianzhang.zj/Code/tidb/executor/explain.go:53\ngithub.com/pingcap/tidb/executor.Next\n\t/Users/jianzhang.zj/Code/tidb/executor/executor.go:249\ngithub.com/pingcap/tidb/executor.(*recordSet).Next\n\t/Users/jianzhang.zj/Code/tidb/executor/adapter.go:126\ngithub.com/pingcap/tidb/server.(*tidbResultSet).Next\n\t/Users/jianzhang.zj/Code/tidb/server/driver_tidb.go:369\ngithub.com/pingcap/tidb/server.(*clientConn).writeChunks\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:1436\ngithub.com/pingcap/tidb/server.(*clientConn).writeResultset\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:1402\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:1309\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:1278\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:901\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/Users/jianzhang.zj/Code/tidb/server/conn.go:715\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/Users/jianzhang.zj/Code/tidb/server/server.go:415"]
```

### 2. What did you expect to see?

no panic, executes successfully.

### 3. What did you see instead?

the error and panic described in the above.

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

```sh
➜ ./bin/tidb-server -V
Release Version: v4.0.0-beta.2-352-g5ede18f10
Git Commit Hash: 5ede18f10eedfe2e3690d7728dec3ffa4b0af2d5
Git Branch: master
UTC Build Time: 2020-04-25 08:41:26
GoVersion: go1.13.10
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

--- a/executor/explainfor_test.go
+++ b/executor/explainfor_test.go
@@ -17,6 +17,8 @@ import (
 	"crypto/tls"
 	"fmt"
 	"math"
+	"strconv"
+	"sync"
 
 	. "github.com/pingcap/check"
 	"github.com/pingcap/parser/auth"
@@ -128,35 +130,68 @@ type testPrepareSerialSuite struct {
 }
 
 func (s *testPrepareSerialSuite) TestExplainForConnPlanCache(c *C) {
-	tk := testkit.NewTestKit(c, s.store)
 	orgEnable := core.PreparedPlanCacheEnabled()
 	defer func() {
 		core.SetPreparedPlanCache(orgEnable)
 	}()
 	core.SetPreparedPlanCache(true)
+
 	var err error
-	tk.Se, err = session.CreateSession4TestWithOpt(s.store, &session.Opt{
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.Se, err = session.CreateSession4TestWithOpt(s.store, &session.Opt{
 		PreparedPlanCache: kvcache.NewSimpleLRUCache(100, 0.1, math.MaxUint64),
 	})
 	c.Assert(err, IsNil)
+	tk2 := testkit.NewTestKitWithInit(c, s.store)
 
-	tk.MustExec("use test")
-	tk.MustExec("drop table if exists t")
-	tk.MustExec("create table t(a int)")
-	rows := tk.MustQuery("select connection_id()").Rows()
-	c.Assert(len(rows), Equals, 1)
-	connID := rows[0][0].(string)
-	tk.MustExec("prepare stmt from 'select * from t where a = ?'")
-	tk.MustExec("set @p0='1'")
-	tk.MustExec("execute stmt using @p0")
-	tkProcess := tk.Se.ShowProcess()
-	ps := []*util.ProcessInfo{tkProcess}
-	tk.Se.SetSessionManager(&mockSessionManager1{PS: ps})
-	tk.MustQuery(fmt.Sprintf("explain for connection %s", connID)).Check(testkit.Rows(
+	tk1.MustExec("use test")
+	tk1.MustExec("drop table if exists t")
+	tk1.MustExec("create table t(a int)")
+	tk1.MustExec("prepare stmt from 'select * from t where a = ?'")
+	tk1.MustExec("set @p0='1'")
+
+	executeQuery := "execute stmt using @p0"
+	explainQuery := "explain for connection " + strconv.FormatUint(tk1.Se.ShowProcess().ID, 10)
+	explainResult := testkit.Rows(
 		"TableReader_7 8000.00 root  data:Selection_6",
 		"└─Selection_6 8000.00 cop[tikv]  eq(cast(test.t.a), 1)",
 		"  └─TableFullScan_5 10000.00 cop[tikv] table:t keep order:false, stats:pseudo",
-	))
+	)
+
+	// Now the ProcessInfo held by mockSessionManager1 will not be updated in real time.
+	// So it needs to be reset every time before tk2 query.
+	// TODO: replace mockSessionManager1 with another mockSessionManager.
+
+	// single test
+	tk1.MustExec(executeQuery)
+	tk2.Se.SetSessionManager(&mockSessionManager1{
+		PS: []*util.ProcessInfo{tk1.Se.ShowProcess()},
+	})
+	tk2.MustQuery(explainQuery).Check(explainResult)
+
+	// multiple test, '1000' is both effective and efficient.
+	repeats := 1000
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		for i := 0; i < repeats; i++ {
+			tk1.MustExec(executeQuery)
+		}
+		wg.Done()
+	}()
+
+	go func() {
+		for i := 0; i < repeats; i++ {
+			tk2.Se.SetSessionManager(&mockSessionManager1{
+				PS: []*util.ProcessInfo{tk1.Se.ShowProcess()},
+			})
+			tk2.MustQuery(explainQuery).Check(explainResult)
+		}
+		wg.Done()
+	}()
+
+	wg.Wait()
 }
 
 func (s *testPrepareSerialSuite) TestExplainDotForExplainPlan(c *C) {

--- a/expression/constant.go
+++ b/expression/constant.go
@@ -102,11 +102,11 @@ func (c *Constant) Clone() Expression {
 
 // GetType implements Expression interface.
 func (c *Constant) GetType() *types.FieldType {
-	if p := c.ParamMarker; p != nil && !p.ctx.GetSessionVars().StmtCtx.InExplainStmt {
+	if c.ParamMarker != nil {
 		// GetType() may be called in multi-threaded context, e.g, in building inner executors of IndexJoin,
 		// so it should avoid data race. We achieve this by returning different FieldType pointer for each call.
 		tp := types.NewFieldType(mysql.TypeUnspecified)
-		dt := p.GetUserVar()
+		dt := c.ParamMarker.GetUserVar()
 		types.DefaultParamTypeForValue(dt.GetValue(), tp)
 		return tp
 	}
@@ -170,21 +170,13 @@ func (c *Constant) VecEvalJSON(ctx sessionctx.Context, input *chunk.Chunk, resul
 }
 
 func (c *Constant) getLazyDatum(row chunk.Row) (dt types.Datum, isLazy bool, err error) {
-	if p := c.ParamMarker; p != nil {
-		if p.ctx.GetSessionVars().StmtCtx.InExplainStmt {
-			// Since `ParamMarker` is not nil only in prepare/execute context, the query must be `explain for connection` when coming here.
-			// The PreparedParams may have been reset already, to avoid panic, we just use the pre-evaluated datum for this constant.
-			return dt, false, nil
-		}
-		dt = p.GetUserVar()
-		isLazy = true
-		return
+	if c.ParamMarker != nil {
+		return c.ParamMarker.GetUserVar(), true, nil
 	} else if c.DeferredExpr != nil {
 		dt, err = c.DeferredExpr.Eval(row)
-		isLazy = true
-		return
+		return dt, true, err
 	}
-	return
+	return types.Datum{}, false, nil
 }
 
 // Eval implements Expression interface.

--- a/planner/core/common_plans.go
+++ b/planner/core/common_plans.go
@@ -798,6 +798,19 @@ type Explain struct {
 	explainedPlans map[int]bool
 }
 
+// GetExplainRowsForPlan get explain rows for plan.
+func GetExplainRowsForPlan(plan Plan) (rows [][]string) {
+	explain := &Explain{
+		TargetPlan: plan,
+		Format:     ast.ExplainFormatROW,
+		Analyze:    false,
+	}
+	if err := explain.RenderResult(); err != nil {
+		return rows
+	}
+	return explain.Rows
+}
+
 // prepareSchema prepares explain's result schema.
 func (e *Explain) prepareSchema() error {
 	var fieldNames []string
@@ -836,10 +849,12 @@ func (e *Explain) RenderResult() error {
 	}
 	switch strings.ToLower(e.Format) {
 	case ast.ExplainFormatROW:
-		e.explainedPlans = map[int]bool{}
-		err := e.explainPlanInRowFormat(e.TargetPlan, "root", "", "", true)
-		if err != nil {
-			return err
+		if e.Rows == nil || e.Analyze {
+			e.explainedPlans = map[int]bool{}
+			err := e.explainPlanInRowFormat(e.TargetPlan, "root", "", "", true)
+			if err != nil {
+				return err
+			}
 		}
 	case ast.ExplainFormatDOT:
 		if physicalPlan, ok := e.TargetPlan.(PhysicalPlan); ok {

--- a/planner/core/planbuilder.go
+++ b/planner/core/planbuilder.go
@@ -2855,12 +2855,13 @@ func (b *PlanBuilder) buildTrace(trace *ast.TraceStmt) (Plan, error) {
 	return p, nil
 }
 
-func (b *PlanBuilder) buildExplainPlan(targetPlan Plan, format string, analyze bool, execStmt ast.StmtNode) (Plan, error) {
+func (b *PlanBuilder) buildExplainPlan(targetPlan Plan, format string, rows [][]string, analyze bool, execStmt ast.StmtNode) (Plan, error) {
 	p := &Explain{
 		TargetPlan: targetPlan,
 		Format:     format,
 		Analyze:    analyze,
 		ExecStmt:   execStmt,
+		Rows:       rows,
 	}
 	p.ctx = b.ctx
 	return p, p.prepareSchema()
@@ -2886,7 +2887,7 @@ func (b *PlanBuilder) buildExplainFor(explainFor *ast.ExplainForStmt) (Plan, err
 		return &Explain{Format: explainFor.Format}, nil
 	}
 
-	return b.buildExplainPlan(targetPlan, explainFor.Format, false, nil)
+	return b.buildExplainPlan(targetPlan, explainFor.Format, processInfo.PlanExplainRows, false, nil)
 }
 
 func (b *PlanBuilder) buildExplain(ctx context.Context, explain *ast.ExplainStmt) (Plan, error) {
@@ -2898,7 +2899,7 @@ func (b *PlanBuilder) buildExplain(ctx context.Context, explain *ast.ExplainStmt
 		return nil, err
 	}
 
-	return b.buildExplainPlan(targetPlan, explain.Format, explain.Analyze, explain.Stmt)
+	return b.buildExplainPlan(targetPlan, explain.Format, nil, explain.Analyze, explain.Stmt)
 }
 
 func (b *PlanBuilder) buildSelectInto(ctx context.Context, sel *ast.SelectStmt) (Plan, error) {

--- a/session/session.go
+++ b/session/session.go
@@ -1017,6 +1017,7 @@ func (s *session) SetProcessInfo(sql string, t time.Time, command byte, maxExecu
 		DB:               s.sessionVars.CurrentDB,
 		Command:          command,
 		Plan:             s.currentPlan,
+		PlanExplainRows:  plannercore.GetExplainRowsForPlan(s.currentPlan),
 		Time:             t,
 		State:            s.Status(),
 		Info:             sql,

--- a/util/processinfo.go
+++ b/util/processinfo.go
@@ -25,16 +25,17 @@ import (
 
 // ProcessInfo is a struct used for show processlist statement.
 type ProcessInfo struct {
-	ID            uint64
-	User          string
-	Host          string
-	DB            string
-	Plan          interface{}
-	Time          time.Time
-	Info          string
-	CurTxnStartTS uint64
-	StmtCtx       *stmtctx.StatementContext
-	StatsInfo     func(interface{}) map[string]uint64
+	ID              uint64
+	User            string
+	Host            string
+	DB              string
+	Plan            interface{}
+	PlanExplainRows [][]string
+	Time            time.Time
+	Info            string
+	CurTxnStartTS   uint64
+	StmtCtx         *stmtctx.StatementContext
+	StatsInfo       func(interface{}) map[string]uint64
 	// MaxExecutionTime is the timeout for select statement, in milliseconds.
 	// If the query takes too long, kill it.
 	MaxExecutionTime uint64

